Finally cleanup (#592)
Vineflower/vineflower: docs/finally-processing.md

# Finally processing

> This document aims to explain how finally's are detected and reconstructed.

Reviewer: ```suggestion
This document aims to explain how `finally` blocks are detected and reconstructed.
```

  `continue` or `yield`). In this case, the execution will not resume where the `try` (or `catch`) block wanted to go
  to, but instead, it will go wherever this "abrupt end" wants us to go to. Method exit in case of `return` or an
  uncaught error, a `catch` block in case of a caught error, the end of a loop in case of `break`, the start of a loop
  in case of a `continue`, the result of a `switch` expression in case of `yield`.

The example code above will thus never `return` and instead run in an infinite loop. It will calculate the result of
`x + 5`, and store it in a temporary variable while it evaluates the `finally` block. But the `finally` block will
always end with the `continue` and thus always restart the loop and ignore the fact that there was a `return`.

This behavior is almost never the intention of the user. This construct can occur in many languages (Java, Python,
JavaScript, ...). Some languages produce warnings, other give compile errors, and those that don't, have linters that
don't allow it.

## How are `finally`s compiled

(This section assumes the Vineflower simplification listed above)

> Before Java 6, `finally`s where compiled using the `jsr` and `ret` opcodes. This allowed for the creation of

Reviewer: ```suggestion
Before Java 6, `finally`s were compiled using the `jsr` and `ret` opcodes. This allowed for the creation of
```
Also, this is technically not true. Javac only used jsr/ret up until java 2ish, until java 6 it was allowed but the official tools didn't emit them.

Author: Ah, I'll change that

mini-functions (subroutines) inside the larger function. At each point in the `try` block, right before an exit out of
that block (`return`, `break`, ... or just normally exiting) the subroutine with the code of the `finally` block was
called. Additionally, there is an exception handler present that catches all exceptions (VF: "catch-all") in the code
for the `try` block which jumps to a piece of code that stores the exception in a variable, also jumps to the `finally`
subroutine and then rethrows the caught exception.

However, even though there are restrictions on the subroutines (no recursing, each subroutine can have at most one `ret`
instruction, and a nested subroutine call can no longer be returned from if an outer sub routine has already returned),
it made validating and optimizing the bytecode non-trivial. As a result, starting in Java 6, and required for classes

> with classfile version 51.0 and above (Java 7), `jsr` and `ret` where no longer allowed and instead, each exit point and

Reviewer: ```suggestion
with classfile version 51.0 and above (Java 7), `jsr` and `ret` were no longer allowed and instead, each exit point and
```

the catch-all will jump to their own copied version of the finally code.

Handling `jsr` and `ret` instructions isn't only annoying for the Java Runtime, it's also annoying for Vineflower.
As a result, one of the very first processing steps on the parsed Control Flow Graph is the inlining of jsr
instructions. This means that we convert old Java 5 and before `finally`s into the version that we'd expect for Java 6+.
Given that in a later step we will need to deduplicate these, this might seem like a weird action to take but abusing

> `jsr` and `ret` where really popular in obfuscators. This meant that assuming that these subroutines could always be

Reviewer: ```suggestion
`jsr` and `ret` was really popular in obfuscators. This meant that assuming that these subroutines could always be
```

a type
of "finally exit". Catch handlers on these sample & pattern blocks are also compared. The exception types must match
and the handlers are then also compared if they are inside the finally catch handler.

The caveats:

(1): In case of the very first block of the finally catch handler, an initial `pop` (discarding the exception) or
`astore` (storing the exception) instruction will be skipped in the comparison. Additionally, for the "normal" exit
out of the finally block, the `throw` (and previous `aload` in case of a `STORE` finally type) are also skipped.

(2): It is permitted for a sample block to contain more instruction than its corresponding pattern block. In this case,
the remaining unmatched instructions will be split into a new basic block, taking over all the successors, copying the
exception handlers and becoming the sole successor of the original sample block

(3): Matching instructions ignores a few things. First, variables defined inside a `finally` block can have different
variable indices between copies. The processor tries to keep a list of match ups but is actually too lenient. Jump
targets in jump and switch blocks are ignored.

> If everything looks oke, then the info is bundled into an "Area" object. If all the jumps out of the try block have been

Reviewer: ```suggestion
If everything looks okay, then the info is bundled into an "Area" object. If all the jumps out of the try block have been
```

validated, the Areas are deleted from the graph, and the exception handling instructions are removed from the finally

> catch handler so it can be used as a propper finally block instead.

Reviewer: ```suggestion
catch handler so it can be used as a proper finally block instead.
```
